### PR TITLE
feat: allow sharding for auraflow.

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

AuraFlow is one of the largest open model right now. A 5.6B parameter model maybe very hard to run on consumer GPUs having 15GBs if we are not careful. 

This PR adds support to run AuraFlow Transformer in 15GBs by allowing offloading at the model level. 

Example code:  

```python
from diffusers import AuraFlowTransformer2DModel
import tempfile
import torch
import os

def get_inputs():
    sample = torch.randn(1, 4, 128, 128)
    timestep = torch.randint(0, 1000, size=(1, ))
    encoder_hidden_states = torch.randn(1, 256, 2048)

    return sample, timestep, encoder_hidden_states

with torch.no_grad():
    max_memory = {0: "15GB"} # reasonable estimate for a consumer-gpu.
    with tempfile.TemporaryDirectory() as tmp_dir:
        new_model = AuraFlowTransformer2DModel.from_pretrained(
            "fal/AuraFlow", 
            subfolder="transformer",
            device_map="auto",
            max_memory=max_memory, 
            offload_folder=os.path.join("/raid/.cache/huggingface", tmp_dir)
        )

        sample, timestep, encoder_hidden_states = get_inputs()
        out = new_model(
            hidden_states=sample,
            encoder_hidden_states=encoder_hidden_states,
            timestep=timestep,
        ).sample
        print(out.shape)
```

I have run the offloading tests and they pass too:

```bash
pytest tests/models/transformers/test_models_pixart_transformer2d.py -k "offload"
```

@SunMarc FYI.

### CODE REVIEW

**Comment by sayakpaul** [comment]
@DN6 a gentle ping.

File: src/diffusers/models/transformers/auraflow_transformer_2d.py [hunk 1]
```diff
@@ -253,6 +253,7 @@ class AuraFlowTransformer2DModel(ModelMixin, ConfigMixin):
         pos_embed_max_size (`int`, defaults to 4096): Maximum positions to embed from the image latents.
     """
 
+    _no_split_modules = ["AuraFlowJointTransformerBlock", "AuraFlowSingleTransformerBlock", "AuraFlowPatchEmbed"]
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/auraflow_transformer_2d.py:256]: I'm curious what's expected behavior of `_no_split_modules` here? 

I assume we would not split these specified layers across different devices, did I understand wrong? 

I think we are splitting the `AuraFlowSingleTransformerBlock` based on the device map  cc @SunMarc 

```
 hf_device_map:
{'register_tokens': 0, 'pos_embed': 0, 'context_embedder': 0, 'time_step_embed': 0, 'time_step_proj': 0, 
'joint_transformer_blocks': 0, 'single_transformer_blocks.0': 0, 'single_transformer_blocks.1': 0, 'single_transformer_blocks.2': 0, 'single_transformer_blocks.3': 0, 'single_transformer_blocks.4': 0, 
'single_transformer_blocks.5': 0, 'single_transformer_blocks.6': 0, 'single_transformer_blocks.7': 0, 
'single_transformer_blocks.8': 0, 'single_transformer_blocks.9': 0, 'single_transformer_blocks.10': 0, 
'single_transformer_blocks.11': 0, 'single_transformer_blocks.12': 'disk', 'single_transformer_blocks.13': 'disk', 'single_transformer_blocks.14': 'disk', 'single_transformer_blocks.15': 'disk', 
'single_transformer_blocks.16': 'disk', 'single_transformer_blocks.17': 'disk', 
'single_transformer_blocks.18': 'disk', 'single_transformer_blocks.19': 'disk', 
'single_transformer_blocks.20': 'disk', 'single_transformer_blocks.21': 'disk', 
'single_transformer_blocks.22': 'disk', 'single_transformer_blocks.23': 'disk', 
'single_transformer_blocks.24': 'disk', 'single_transformer_blocks.25': 'disk', 
'single_transformer_blocks.26': 'disk', 'single_transformer_blocks.27': 'disk', 
'single_transformer_blocks.28': 'disk', 'single_transformer_blocks.29': 'disk', 
'single_transformer_blocks.30': 'disk', 'single_transformer_blocks.31': 'disk',
 'norm_out': 'disk', 'proj_out': 'disk'}
```

Comment (sayakpaul) [reply src/diffusers/models/transformers/auraflow_transformer_2d.py:256]: `_no_split_modules` are the modules that contain residual connections. As we can see that the blocks specified here -- all of them contain residual connections. Hence we cannot split. But I am happy to stand corrected @SunMarc.

Comment (sayakpaul) [reply src/diffusers/models/transformers/auraflow_transformer_2d.py:256]: @yiyixuxu I think Marc agrees to what I said here. Could you take another look?

**Comment by sayakpaul** [comment]
@DN6 could you give this a check?

**Review by yiyixuxu — APPROVED** [review]
sorry I missed this one!
thanks!

File: src/diffusers/models/transformers/auraflow_transformer_2d.py (hunk 1 above)
Comment (yiyixuxu) [reply src/diffusers/models/transformers/auraflow_transformer_2d.py:261]: sorry this PR somehow got missed - approved

**Pull request merged**